### ISSUE TITLE
Allow includepattens for individual sniff codes

### ISSUE THREAD

**Issue by gmponos** [squizlabs/PHP_CodeSniffer#1981]
Hello,

I have this part of xml for my rules:

```xml
    <rule ref="Squiz.Strings.DoubleQuoteUsage">
        <exclude name="Squiz.Strings.DoubleQuoteUsage.ContainsVar" />
        <include-pattern>*/myfile.php</include-pattern>
    </rule>
    </rule>
    <rule ref="Squiz.Commenting.InlineComment.SpacingAfter">
        <include-pattern>*/myfile.php</include-pattern>
    </rule>
    <rule ref="Squiz.Commenting.InlineComment.NoSpaceBefore">
        <include-pattern>*/myfile.php</include-pattern>
    </rule>
```

This part works fine:
    <rule ref="Squiz.Strings.DoubleQuoteUsage">
        <exclude name="Squiz.Strings.DoubleQuoteUsage.ContainsVar" />
        <include-pattern>*/myfile.php</include-pattern>
    </rule>

But the rest two parts are checking all the files instead of only checking `myfile.php`

Is this a known issue? How can I help debugging this?

**Comment by gsherwood** [comment]
> How can I help debugging this?

Running PHPCS with `-vv` will show the ruleset parsing output at the top. This would help determine if PHPCS saw the rule and applied it to the right sniffs.

**Comment by gmponos** [comment]
OK I created a new project. From composer we have that it uses the following version `squizlabs/php_codesniffer 3.2.3`

Inside the project I created a new folder `src` with two files:

**testfile1.php**
```
<?php

//mycomment
```

**testfile2.php**
```
<?php

//mycomment
$string = "mystring";
```

The ruleset I used was a file phpcs.xml.dist containing this:
```xml
<?xml version="1.0"?>
<ruleset name="Ruleset">
    <rule ref="Squiz.Strings.DoubleQuoteUsage">
        <exclude name="Squiz.Strings.DoubleQuoteUsage.ContainsVar" />
        <include-pattern>*/src/testfile2.php</include-pattern>
    </rule>
    <rule ref="Squiz.Commenting.InlineComment.SpacingAfter">
        <include-pattern>*/src/testfile2.php</include-pattern>
    </rule>
    <rule ref="Squiz.Commenting.InlineComment.NoSpaceBefore">
        <include-pattern>*/src/testfile2.php</include-pattern>
    </rule>
</ruleset>
```

In other words the rules should apply only to `testfile2.php`

The output of `php vendor/bin/phpcs src/ -vv` is the following:

```
Processing ruleset /testproject/phpcs.xml.dist
	Processing rule "Squiz.Strings.DoubleQuoteUsage"
		=> /testproject/vendor/squizlabs/php_codesniffer/src/Standards/Squiz/Sniffs/Strings/DoubleQuoteUsageSniff.php
		Excluding rule "Squiz.Strings.DoubleQuoteUsage.ContainsVar"
		=> severity set to 0
		=> added rule-specific absolute include pattern: */src/testfile2.php
	Processing rule "Squiz.Commenting.InlineComment.SpacingAfter"
		=> /testproject/vendor/squizlabs/php_codesniffer/src/Standards/Squiz/Sniffs/Commenting/InlineCommentSniff.php
		Excluding sniff "Squiz.Commenting.InlineComment" except for "SpacingAfter"
		=> added rule-specific absolute include pattern: */src/testfile2.php
	Processing rule "Squiz.Commenting.InlineComment.NoSpaceBefore"
		=> /testproject/vendor/squizlabs/php_codesniffer/src/Standards/Squiz/Sniffs/Commenting/InlineCommentSniff.php
		* disabling sniff exclusion for specific message code *
		=> severity set to 5
		=> added rule-specific absolute include pattern: */src/testfile2.php
=> Ruleset processing complete; included 2 sniffs and excluded 0
Creating file list... DONE (2 files in queue)
Changing into directory /testproject/src
Processing testfile1.php 
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG => <?php[30;1m\n[0m
	Process token [1]: T_WHITESPACE => [30;1m\n[0m
	Process token [2]: T_COMMENT => //mycomment
	*** END PHP TOKENIZING ***
	*** START TOKEN MAP ***
	*** END TOKEN MAP ***
	*** START SCOPE MAP ***
	*** END SCOPE MAP ***
	*** START LEVEL MAP ***
	Process token 0 on line 1 [col:1;len:5;lvl:0;]: T_OPEN_TAG => <?php[30;1m\n[0m
	Process token 1 on line 2 [col:1;len:0;lvl:0;]: T_WHITESPACE => [30;1m\n[0m
	Process token 2 on line 3 [col:1;len:11;lvl:0;]: T_COMMENT => //mycomment
	*** END LEVEL MAP ***
	*** START ADDITIONAL PHP PROCESSING ***
	*** END ADDITIONAL PHP PROCESSING ***
[PHP => 3 tokens in 3 lines]... 
DONE in 10ms (1 errors, 0 warnings)
Processing testfile2.php 
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG => <?php[30;1m\n[0m
	Process token [1]: T_WHITESPACE => [30;1m\n[0m
	Process token [2]: T_COMMENT => //mycomment[30;1m\n[0m
	Process token [3]: T_VARIABLE => $string
	Process token [4]: T_WHITESPACE => [30;1m·[0m
	Process token  5 : T_EQUAL => =
	Process token [6]: T_WHITESPACE => [30;1m·[0m
	Process token [7]: T_CONSTANT_ENCAPSED_STRING => "mystring"
	Process token  8 : T_SEMICOLON => ;
	*** END PHP TOKENIZING ***
	*** START TOKEN MAP ***
	*** END TOKEN MAP ***
	*** START SCOPE MAP ***
	*** END SCOPE MAP ***
	*** START LEVEL MAP ***
	Process token 0 on line 1 [col:1;len:5;lvl:0;]: T_OPEN_TAG => <?php[30;1m\n[0m
	Process token 1 on line 2 [col:1;len:0;lvl:0;]: T_WHITESPACE => [30;1m\n[0m
	Process token 2 on line 3 [col:1;len:11;lvl:0;]: T_COMMENT => //mycomment[30;1m\n[0m
	Process token 3 on line 4 [col:1;len:7;lvl:0;]: T_VARIABLE => $string
	Process token 4 on line 4 [col:8;len:1;lvl:0;]: T_WHITESPACE => [30;1m·[0m
	Process token 5 on line 4 [col:9;len:1;lvl:0;]: T_EQUAL => =
	Process token 6 on line 4 [col:10;len:1;lvl:0;]: T_WHITESPACE => [30;1m·[0m
	Process token 7 on line 4 [col:11;len:10;lvl:0;]: T_CONSTANT_ENCAPSED_STRING => "mystring"
	Process token 8 on line 4 [col:21;len:1;lvl:0;]: T_SEMICOLON => ;
	*** END LEVEL MAP ***
	*** START ADDITIONAL PHP PROCESSING ***
	*** END ADDITIONAL PHP PROCESSING ***
[PHP => 9 tokens in 4 lines]... 
DONE in 1ms (2 errors, 0 warnings)

FILE: /testproject/src/testfile1.php
-----------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------------
 3 | ERROR | [x] No space found before comment text; expected "// mycomment" but found "//mycomment"
-----------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-----------------------------------------------------------------------------------------------------


FILE: /testproject/src/testfile2.php
-----------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------------
 3 | ERROR | [x] No space found before comment text; expected "// mycomment" but found "//mycomment"
 4 | ERROR | [x] String "mystring" does not require double quotes; use single quotes instead
-----------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-----------------------------------------------------------------------------------------------------

Time: 41ms; Memory: 2.25Mb
```

I hope this helps.

**Comment by gsherwood** [comment]
Sorry, I missed what you were trying to do in your original comment.

Include and exclude pattens only apply to sniff classes. They don't work for individual sniff codes like the way you are trying to use them.

This would be an entirely new feature and would need a fair chunk of core changes as these patterns are only used to determine what sniffs are given to a file to process. The file itself applies no regular expression checks when it is storing error messages, and I imagine you'd take a pretty decent performance hit if it was.

I'll leave this open as a feature request but I also change the docs to make it clear these tags work with sniff codes and not sniff message codes.

**Comment by gsherwood** [comment]
...and I just remembered that exclude patterns probably work with sniff codes already. So I'll go test that first and see if, given that structure, there is any quick way to add include pattern support as well.

**Comment by gsherwood** [comment]
I can support these pretty easily:
```xml
<rule ref="Squiz.Commenting.InlineComment.SpacingAfter">
    <include-pattern>*/src/testfile2.php</include-pattern>
</rule>
<rule ref="Squiz.Commenting.InlineComment.NoSpaceBefore">
    <include-pattern>*/src/testfile2.php</include-pattern>
</rule>
```
But I want to clarify what this one will do:
```xml
<rule ref="Squiz.Strings.DoubleQuoteUsage">
    <exclude name="Squiz.Strings.DoubleQuoteUsage.ContainsVar" />
    <include-pattern>*/src/testfile2.php</include-pattern>
</rule>
```
This rule will only run the `Squiz.Strings.DoubleQuoteUsage` sniff for files that match the pattern `*/src/testfile2.php` and will exclude the `Squiz.Strings.DoubleQuoteUsage.ContainsVar` message for those files as well. Hopefully that's what you're expecting.

**Comment by gsherwood** [comment]
I've added this feature now, and updated the docs again. It will be released in 3.3.0. Sorry for my confusing initial comment - had a brain explosion obviously.
